### DISCUSSION TITLE
How to profile halide on macos?

### DISCUSSION THREAD

**Discussion by FindHao** [halide/Halide#8409] [Q&A]
Hi, I'm trying to test halide metal backend on apple silicon. Could you tell me the correct way to profile a python testing example using halide python binding? Is there something like nsight system for NVIDIA we can use to profile? Thanks

**Comment by abadams** [comment]
We don't have anything metal-specific. I recommend using the native profiling tools available in the instruments app.

You can also do somewhat crude profiling by adding the target feature "-debug", which adds synchronization to GPU API calls and causes a lot of debugging spew, including how long each kernel launch took.

**Reply by FindHao** [reply]
Thanks for your reply! Maybe a silly question, but do you have any suggestions on which native profiling tools for macos with metal backend? Thanks!

**Reply by abadams** [reply]
I usually just do CPU profiling on macos, but @shoaibkamil would know the answer to that.
